{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Combo</th>\n",
       "      <th>muts</th>\n",
       "      <th>fit</th>\n",
       "      <th>zs score</th>\n",
       "      <th>num_muts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMHG</td>\n",
       "      <td>V39A:D40M:G41H:V54G</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>147.39927</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>QPEI</td>\n",
       "      <td>V39Q:D40P:G41E:V54I</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>53.49178</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GMYW</td>\n",
       "      <td>V39G:D40M:G41Y:V54W</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>134.21443</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KWNA</td>\n",
       "      <td>V39K:D40W:G41N:V54A</td>\n",
       "      <td>0.001791</td>\n",
       "      <td>145.52714</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>QDRA</td>\n",
       "      <td>V39Q:G41R:V54A</td>\n",
       "      <td>0.004730</td>\n",
       "      <td>139.14162</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149356</th>\n",
       "      <td>CCRA</td>\n",
       "      <td>V39C:D40C:G41R:V54A</td>\n",
       "      <td>0.009981</td>\n",
       "      <td>141.02978</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149357</th>\n",
       "      <td>QSSK</td>\n",
       "      <td>V39Q:D40S:G41S:V54K</td>\n",
       "      <td>0.004755</td>\n",
       "      <td>145.35619</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149358</th>\n",
       "      <td>IAFG</td>\n",
       "      <td>V39I:D40A:G41F:V54G</td>\n",
       "      <td>3.028731</td>\n",
       "      <td>149.25890</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149359</th>\n",
       "      <td>TILI</td>\n",
       "      <td>V39T:D40I:G41L:V54I</td>\n",
       "      <td>0.003469</td>\n",
       "      <td>141.88873</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149360</th>\n",
       "      <td>LLPC</td>\n",
       "      <td>V39L:D40L:G41P:V54C</td>\n",
       "      <td>0.008662</td>\n",
       "      <td>76.27056</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>149361 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Combo                 muts       fit   zs score  num_muts\n",
       "0       AMHG  V39A:D40M:G41H:V54G  0.000000  147.39927         4\n",
       "1       QPEI  V39Q:D40P:G41E:V54I  0.000000   53.49178         4\n",
       "2       GMYW  V39G:D40M:G41Y:V54W  0.000000  134.21443         4\n",
       "3       KWNA  V39K:D40W:G41N:V54A  0.001791  145.52714         4\n",
       "4       QDRA       V39Q:G41R:V54A  0.004730  139.14162         3\n",
       "...      ...                  ...       ...        ...       ...\n",
       "149356  CCRA  V39C:D40C:G41R:V54A  0.009981  141.02978         4\n",
       "149357  QSSK  V39Q:D40S:G41S:V54K  0.004755  145.35619         4\n",
       "149358  IAFG  V39I:D40A:G41F:V54G  3.028731  149.25890         4\n",
       "149359  TILI  V39T:D40I:G41L:V54I  0.003469  141.88873         4\n",
       "149360  LLPC  V39L:D40L:G41P:V54C  0.008662   76.27056         4\n",
       "\n",
       "[149361 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv('GB1_fitness.csv')\n",
    "data = data.drop(['EvMutation', 'fit_preds', 'indep_fit'], axis =1)\n",
    "data = data.rename(columns={\"Triad-FixedBb-dG\": \"zs score\"})\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "data.to_csv('GB1_fitness.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DeCOIL\n",
    "DEgnerate Codon Optimization for Informed Libraries (DeCOIL) is a computational method to optimize degnerate codos for synthesizing an informed combinatorial mutagenesis protein variant library, which is a critical step in a variety of machine learning-assisted protein engineering workflows. DeCOIL aims to bias a degenerate library toward desired proteins based on a predictive model from physics-based simulations, evolutionary knowledge, or machine-learning models while also providing diversity in the library.\n",
    "\n",
    "## Installation\n",
    "To download, clone this repository\n",
    "```\n",
    "git clone https://github.com/jsunn-y/DeCOIL.git\n",
    "```\n",
    "To run DeCOIL optimization, the relevant anaconda environment can be installed from `DeCOIL.yml`. To build this environment, run\n",
    "```\n",
    "cd ./DeCOIL\n",
    "conda env create -f DeCOIL.yml\n",
    "conda activate DeCOIL\n",
    "```\n",
    "\n",
    "## General Use\n",
    "DeCOIL can be executed on a personal computer using the following command:\n",
    "```\n",
    "python execute_optimization.py --config_file 'file_name.json'\n",
    "```\n",
    "The configuration file should be located in `MLDE_lite/configs` with keys specified as follows:\n",
    "| Key | Description |\n",
    "|:-------|:----------------:|\n",
    "| name |filname for csv file containing sequences in the combinatorial space and corresponding predicted values. Must contain a column called 'Combo' which sqecifies the mutations at each site in the combinatorial library| \n",
    "|zs_name | Name of the column in the csv file, which contains the relative value of each variant. By default, lower values are more desirable.| \n",
    "|samples |Approximate screening size (number of sequences sampled from the library)| \n",
    "|sites |number of amino acid sites targeted in the combinatoral library|\n",
    "\n",
    "\n",
    "| Key | Description |\n",
    "|:-------|:----------------:|\n",
    "| type |type of optimization, only \"greedy\" is currently supported| \n",
    "|weight_type | List containing either \"simple\" or \"full\" for step and diffuse coverage respectively. If \"full\" is specified, the encoding type can also be specified, default is onehot.| \n",
    "|dist_function |Only required if weight_type contains \"full\". Can be \"hamming\", \"manhattan\", or \"euclidean\" | \n",
    "| zs_exp |p parameter, which is used to tune the desired amount of exploration and exploitation|\n",
    "| sigma | $\\sigma$ parameter used in diffuse coverage calculations|\n",
    "| seed | seed for reproducibility|\n",
    "| iters | number of iterations for greedy hill climbing |\n",
    "| samples | number of initial starting points and solutions, parallelized|\n",
    "| n_mix | number of templates per library to optimize simultaneously |\n",
    "| directions | number of random changes to the library to test in each iteration of greedy hill climbing|\n",
    "| num_repeats | number of sets of sequences to sample from the library for scoring, to get good statistics|\n",
    "| num_workers | number of processors to use|\n",
    "| top_fraction | fraction of sequences in a certain percentile of predictor scores, for logging purposes only|\n",
    "\n",
    "\n",
    "Outputs will be saved in `optimization/saved` under a folder with the name as the configuration file. The optimization trajectory can be accessed in `results.npy` and example analyses are provided in `analysis.ipynb`.\n",
    "\n",
    "## MLDE Simulations\n",
    "The MLDE simulations demonstrated in our accompanying study can be reproduced below:\n",
    "```\n",
    "python execute_optimization.py --config_file 'file_name.json'\n",
    "```\n",
    "The configuration file should be located in `optimization/configs` with keys specified as follows:\n",
    "| Key | Description |\n",
    "|:-------|:----------------:|\n",
    "| name |filname for csv file containing sequences in the combinatorial space and corresponding fitness values. Use \"GB1_fitness.csv\" for the GB1 dataset| \n",
    "|encoding | List of encodings to use, only supports \"one-hot\"| \n",
    "|library | Specifies how the training sequences should be sampled for each simulation. If a list of integers is provided, for an integer $a$, sequences will be sampled from the library containing the top $a$ sequences based on the Triad rank. If a string is specified, the optimized degenerate codon libraries corresponding to that folder in `optimization/saved` will be used| \n",
    "|n_solutions |must correspond to the number of libraries specified in \"library\"|\n",
    "\n",
    "| Key | Description |\n",
    "|:-------|:----------------:|\n",
    "| name | list of model classes to be used. Only supports \"boosting\"| \n",
    "\n",
    "| Key | Description |\n",
    "|:-------|:----------------:|\n",
    "| seed | seed for reproducibility| \n",
    "| n_samples | list of integers specifying the number of sequences to sample from a library for the training set| \n",
    "| n_splits | the number of models to be trained in the ensemble| \n",
    "| n_subsets | number of times to sample training sequences from the library, to get statistics| \n",
    "| num_workers | number of processors to use| \n",
    "| verbose | boolean specifying whether or not to print training updates| \n",
    "| save_model | boolean specifying whether or not to save the trained models| \n",
    "\n",
    "Outputs will be saved in `MLDE_lite/saved` under a folder with the name as the configuration file. The results of MLDE simulations can be accessed in `mlde_results.npy` and example analyses are provided in `analysis.ipynb`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ProtGraphR",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "139d1aaa1d1fb05ac2b871ea076ee69809b0d9a8dfdac7dc2e2d4e05ffa29f01"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
